plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/file/TestIcebergFileCreateTableFailureTest.java

```diff
@@ -0,0 +1,139 @@
+                HDFS_ENVIRONMENT,
+                new HiveMetastoreConfig().isHideDeltaLakeTables(),
+                new FileHiveMetastoreConfig()
+                        .setCatalogDirectory(dataDirectory.toString())
+                        .setMetastoreUser("test"))
+        {
+            @Override
+            public synchronized void createTable(Table table, PrincipalPrivileges principalPrivileges)
+            {
+                throw testException.get();
+            }
+        };
+
+        this.fileSystem = new HdfsFileSystemFactory(HDFS_ENVIRONMENT).create(TestingConnectorSession.SESSION);
+
+        Session session = testSessionBuilder()
+                .setCatalog(ICEBERG_CATALOG)
+                .setSchema(SCHEMA_NAME)
+                .build();
+
+        DistributedQueryRunner queryRunner = DistributedQueryRunner.builder(session).build();
+        queryRunner.installPlugin(new TestingIcebergPlugin(Optional.of(new TestingIcebergFileMetastoreCatalogModule(metastore)), Optional.empty(), EMPTY_MODULE));
+        queryRunner.createCatalog("iceberg", "iceberg");
+        queryRunner.execute("CREATE SCHEMA " + SCHEMA_NAME);
+
+        return queryRunner;
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void cleanup()
+            throws Exception
+    {
+        if (metastore != null) {
+            metastore.dropDatabase(SCHEMA_NAME, true);
+        }
+        if (dataDirectory != null) {
+            deleteRecursively(dataDirectory, ALLOW_INSECURE);
+        }
+    }
+
+    @Test
+    public void testCreateTableFailureMetadataCleanedUp()
+            throws Exception
+    {
+        final String exceptionMessage = "Test-simulated metastore schema not found exception";
+        testException.set(new SchemaNotFoundException(getSession().getSchema().orElseThrow(), exceptionMessage));
+        testCreateTableFailure(exceptionMessage, false);
+    }
+
+    @Test
+    public void testCreateTableFailureMetadataNotCleanedUp()
+            throws Exception
+    {
+        final String exceptionMessage = "Test-simulated metastore runtime exception";
+        testException.set(new RuntimeException(exceptionMessage));
+        testCreateTableFailure(exceptionMessage, true);
+    }
+
+    protected void testCreateTableFailure(String expectedExceptionMessage, boolean shouldMetadataFileExist)
+            throws IOException
+    {
+        String tableName = "test_create_failure_" + randomNameSuffix();
+        String tableLocation = Path.of(dataDirectory.toString(), tableName).toString();
+        assertThatThrownBy(() -> getQueryRunner().execute(format("CREATE TABLE " + tableName + " (a varchar) with (location = '" + tableLocation + "')")))
+                .hasMessageContaining(expectedExceptionMessage);
+
+        assertThat(fileSystem.listFiles(Path.of(tableLocation, "metadata").toString()).hasNext())
+                .as("Metadata file should" + (shouldMetadataFileExist ? "" : "not") + " exist").isEqualTo(shouldMetadataFileExist);
```

Comment:
The variable name is shouldMetadata FileExist (**_file exist_**), but this check verifies only whether the directory is empty or not. It would be nice to rename the variable or change the assertion logic. e.g.
```java
        Path metadataDirectory = Path.of(tableLocation, "metadata");
        if (shouldMetadataFileExist) {
            assertThat(metadataDirectory).isDirectoryContaining("glob:**.metadata.json");
        }
        else {
            assertThat(metadataDirectory).isEmptyDirectory();
        }
```